File: packages/luciq_flutter/android/src/main/java/ai/luciq/flutter/modules/SessionReplayApi.java
package ai.luciq.flutter.modules;

import androidx.annotation.NonNull;
import ai.luciq.flutter.generated.SessionReplayPigeon;
import ai.luciq.flutter.util.ArgsRegistry;
import ai.luciq.flutter.util.LuciqFlutterDebugTags;
import ai.luciq.flutter.util.LuciqFlutterLogger;
import ai.luciq.library.sessionreplay.SessionReplay;
import io.flutter.plugin.common.BinaryMessenger;

public class SessionReplayApi implements SessionReplayPigeon.SessionReplayHostApi {

    public static void init(BinaryMessenger messenger) {
        final SessionReplayApi api = new SessionReplayApi();
        SessionReplayPigeon.SessionReplayHostApi.setup(messenger, api);
    }

    @Override
    public void setEnabled(@NonNull Boolean isEnabled) {
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setEnabled] phase=enter isEnabled=" + isEnabled);
        SessionReplay.setEnabled(isEnabled);
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setEnabled] phase=exit");
    }

    @Override
    public void setNetworkLogsEnabled(@NonNull Boolean isEnabled) {
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setNetworkLogsEnabled] phase=enter isEnabled=" + isEnabled);
        SessionReplay.setNetworkLogsEnabled(isEnabled);
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setNetworkLogsEnabled] phase=exit");
    }

    @Override
    public void setLuciqLogsEnabled(@NonNull Boolean isEnabled) {
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setLuciqLogsEnabled] phase=enter isEnabled=" + isEnabled);
        SessionReplay.setLuciqLogsEnabled(isEnabled);
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setLuciqLogsEnabled] phase=exit");
    }

    @Override
    public void setUserStepsEnabled(@NonNull Boolean isEnabled) {
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setUserStepsEnabled] phase=enter isEnabled=" + isEnabled);
        SessionReplay.setUserStepsEnabled(isEnabled);
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setUserStepsEnabled] phase=exit");
    }

    @Override
    public void getSessionReplayLink(@NonNull String callId, @NonNull SessionReplayPigeon.Result<String> result) {
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.getSessionReplayLink] #" + callId + " phase=enter");
        SessionReplay.getSessionReplayLink(link -> {
            LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                    "[SR.getSessionReplayLink] #" + callId + " phase=exit resultLength="
                            + (link != null ? link.length() : 0)
                            + " resultPresent=" + (link != null && !link.isEmpty()));
            result.success(link);
        });
    }

    @Override
    public void setScreenshotCapturingMode(@NonNull String mode) {
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setScreenshotCapturingMode] phase=enter mode=" + mode);
        final Integer capturingMode = ArgsRegistry.screenshotCapturingModes.get(mode);
        if (capturingMode == null) {
            LuciqFlutterLogger.e(LuciqFlutterDebugTags.SESSION_REPLAY,
                    "[SR.setScreenshotCapturingMode] phase=error errorType=UnknownEnum mode=" + mode);
            return;
        }
        SessionReplay.setCapturingMode(capturingMode);
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setScreenshotCapturingMode] phase=exit");
    }

    @Override
    public void setScreenshotCaptureInterval(@NonNull Long intervalMs) {
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setScreenshotCaptureInterval] phase=enter intervalMs=" + intervalMs);
        if (intervalMs < 500L) {
            LuciqFlutterLogger.e(LuciqFlutterDebugTags.SESSION_REPLAY,
                    "[SR.setScreenshotCaptureInterval] phase=error errorType=InvalidArgument intervalMs=" + intervalMs);
            return;
        }
        SessionReplay.setScreenshotCaptureInterval(intervalMs.intValue());
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setScreenshotCaptureInterval] phase=exit");
    }

    @Override
    public void setScreenshotQualityMode(@NonNull String mode) {
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setScreenshotQualityMode] phase=enter mode=" + mode);
        final Integer quality = ArgsRegistry.screenshotQualityModes.get(mode);
        if (quality == null) {
            LuciqFlutterLogger.e(LuciqFlutterDebugTags.SESSION_REPLAY,
                    "[SR.setScreenshotQualityMode] phase=error errorType=UnknownEnum mode=" + mode);
            return;
        }
        SessionReplay.setScreenshotQuality(quality);
        LuciqFlutterLogger.d(LuciqFlutterDebugTags.SESSION_REPLAY,
                "[SR.setScreenshotQualityMode] phase=exit");
    }

}

File: packages/luciq_flutter/ios/luciq_flutter/Sources/luciq_flutter/Modules/SessionReplayApi.m
#import <Flutter/Flutter.h>
#import <LuciqSDK/LuciqSDK.h>
#import <LuciqSDK/LCQSessionReplay.h>
#import "SessionReplayApi.h"
#import "ArgsRegistry.h"
#import "../Util/LuciqFlutterLogger.h"
#import "../Util/LuciqFlutterDebugTags.h"

extern void InitSessionReplayApi(id<FlutterBinaryMessenger> messenger) {
    SessionReplayApi *api = [[SessionReplayApi alloc] init];
    SessionReplayHostApiSetup(messenger, api);
}

@implementation SessionReplayApi

- (void)setEnabledIsEnabled:(nonnull NSNumber *)isEnabled error:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay]
                   format:@"[SR.setEnabled] phase=enter isEnabled=%@",
        ([isEnabled boolValue] ? @"true" : @"false")];
    LCQSessionReplay.enabled = [isEnabled boolValue];
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay] format:@"[SR.setEnabled] phase=exit"];
}

- (void)setLuciqLogsEnabledIsEnabled:(nonnull NSNumber *)isEnabled error:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay]
                   format:@"[SR.setLuciqLogsEnabled] phase=enter isEnabled=%@",
        ([isEnabled boolValue] ? @"true" : @"false")];
    LCQSessionReplay.LCQLogsEnabled = [isEnabled boolValue];
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay] format:@"[SR.setLuciqLogsEnabled] phase=exit"];
}

- (void)setNetworkLogsEnabledIsEnabled:(nonnull NSNumber *)isEnabled error:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay]
                   format:@"[SR.setNetworkLogsEnabled] phase=enter isEnabled=%@",
        ([isEnabled boolValue] ? @"true" : @"false")];
    LCQSessionReplay.networkLogsEnabled = [isEnabled boolValue];
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay] format:@"[SR.setNetworkLogsEnabled] phase=exit"];
}

- (void)setUserStepsEnabledIsEnabled:(nonnull NSNumber *)isEnabled error:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay]
                   format:@"[SR.setUserStepsEnabled] phase=enter isEnabled=%@",
        ([isEnabled boolValue] ? @"true" : @"false")];
    LCQSessionReplay.userStepsEnabled = [isEnabled boolValue];
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay] format:@"[SR.setUserStepsEnabled] phase=exit"];
}

- (void)getSessionReplayLinkCallId:(NSString *)callId completion:(void (^)(NSString *, FlutterError *))completion {
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay]
                   format:@"[SR.getSessionReplayLink] #%@ phase=enter", callId];
    NSString *link = LCQSessionReplay.sessionReplayLink;
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay]
                   format:@"[SR.getSessionReplayLink] #%@ phase=exit resultLength=%lu resultPresent=%@",
        callId,
        (unsigned long)link.length,
        (link.length > 0 ? @"true" : @"false")];
    completion(link, nil);
}

- (void)setScreenshotCapturingModeMode:(NSString *)mode error:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay]
                   format:@"[SR.setScreenshotCapturingMode] phase=enter mode=%@", mode];
    LCQScreenshotCapturingMode nativeMode = (ArgsRegistry.screenshotCapturingModes[mode]).integerValue;
    if (ArgsRegistry.screenshotCapturingModes[mode] == nil) {
        [LuciqFlutterLogger w:[LuciqFlutterDebugTags sessionReplay]
                       format:@"[SR.setScreenshotCapturingMode] phase=warn errorType=UnknownEnum mode=%@", mode];
    }
    LCQSessionReplay.screenshotCapturingMode = nativeMode;
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay] format:@"[SR.setScreenshotCapturingMode] phase=exit"];
}

- (void)setScreenshotCaptureIntervalIntervalMs:(nonnull NSNumber *)intervalMs error:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay]
                   format:@"[SR.setScreenshotCaptureInterval] phase=enter intervalMs=%@", intervalMs];
    if ([intervalMs integerValue] < 500) {
        [LuciqFlutterLogger e:[LuciqFlutterDebugTags sessionReplay]
                       format:@"[SR.setScreenshotCaptureInterval] phase=error errorType=InvalidArgument errorMessage=intervalMs<500 intervalMs=%@", intervalMs];
        *error = [FlutterError errorWithCode:@"INVALID_CAPTURE_INTERVAL"
                                     message:@"intervalMs must be >= 500 on iOS"
                                     details:intervalMs];
        return;
    }
    LCQSessionReplay.screenshotCaptureInterval = [intervalMs integerValue];
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay] format:@"[SR.setScreenshotCaptureInterval] phase=exit"];
}

- (void)setScreenshotQualityModeMode:(NSString *)mode error:(FlutterError * _Nullable __autoreleasing * _Nonnull)error {
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay]
                   format:@"[SR.setScreenshotQualityMode] phase=enter mode=%@", mode];
    LCQScreenshotQualityMode nativeMode = (ArgsRegistry.screenshotQualityModes[mode]).integerValue;
    if (ArgsRegistry.screenshotQualityModes[mode] == nil) {
        [LuciqFlutterLogger w:[LuciqFlutterDebugTags sessionReplay]
                       format:@"[SR.setScreenshotQualityMode] phase=warn errorType=UnknownEnum mode=%@", mode];
    }
    LCQSessionReplay.screenshotQualityMode = nativeMode;
    [LuciqFlutterLogger d:[LuciqFlutterDebugTags sessionReplay] format:@"[SR.setScreenshotQualityMode] phase=exit"];
}

@end
